File: core/jvm/src/main/java/dev/tauri/choam/internal/refs/RefP1P1.java
/*
 * SPDX-License-Identifier: Apache-2.0
 * Copyright 2016-2025 Daniel Urban and contributors listed in NOTICE.txt
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package dev.tauri.choam.internal.refs;

import java.lang.invoke.MethodHandles;
import java.lang.invoke.VarHandle;
import java.lang.ref.WeakReference;

import dev.tauri.choam.internal.VarHandleHelper;
import dev.tauri.choam.internal.mcas.Version;
import dev.tauri.choam.core.Ref;

final class RefP1P1<A, B> extends PaddingForP1P1<A, B> implements Ref2Impl<A, B> {

  private static final VarHandle VALUE_B;
  private static final VarHandle VERSION_B;
  private static final VarHandle MARKER_B;

  static {
    try {
      MethodHandles.Lookup l = MethodHandles.lookup();
      VALUE_B = VarHandleHelper.withInvokeExactBehavior(l.findVarHandle(RefP1P1.class, "valueB", Object.class));
      VERSION_B = VarHandleHelper.withInvokeExactBehavior(l.findVarHandle(RefP1P1.class, "versionB", long.class));
      MARKER_B = VarHandleHelper.withInvokeExactBehavior(l.findVarHandle(RefP1P1.class, "markerB", WeakReference.class));
    } catch (ReflectiveOperationException e) {
      throw new ExceptionInInitializerError(e);
    }
  }

  private final long _id1;

  private volatile B valueB;

  private volatile long versionB = Version.Start;

  private volatile WeakReference<Object> markerB; // = null

  private final Ref<B> refB = new Ref2Ref2<A, B>(this);

  public RefP1P1(A a, B b, long i0, long i1) {
    super(a, i0);
    this._id1 = i1;
    this.valueB = b;
  }

  public RefP1P1(long i0, long i1) {
    super(i0);
    this._id1 = i1;
  }

  @Override
  protected final String refToString() {
    return package$.MODULE$.refStringFrom8Ids(this.id0(), this.id1());
  }

  @Override
  public final Ref<B> _2() {
    return this.refB;
  }

  @Override
  public final B unsafeGet2V() {
    return this.valueB;
  }

  @Override
  public final B unsafeGet2P() {
    return (B) VALUE_B.get(this);
  }

  @Override
  public final void unsafeSet2V(B b) {
    this.valueB = b;
  }

  @Override
  public final void unsafeSet2P(B b) {
    VALUE_B.set(this, b);
  }

  @Override
  public final boolean unsafeCas2V(B ov, B nv) {
    return VALUE_B.compareAndSet(this, ov, nv);
  }

  @Override
  public final B unsafeCmpxchg2V(B ov, B nv) {
    return (B) VALUE_B.compareAndExchange(this, ov, nv);
  }

  @Override
  public final B unsafeCmpxchg2R(B ov, B nv) {
    return (B) VALUE_B.compareAndExchangeRelease(this, ov, nv);
  }

  @Override
  public final long unsafeGetVersion2V() {
    return this.versionB;
  }

  @Override
  public final long unsafeCmpxchgVersion2V(long ov, long nv) {
    return (long) VERSION_B.compareAndExchange(this, ov, nv);
  }

  @Override
  public final WeakReference<Object> unsafeGetMarker2V() {
    return this.markerB;
  }

  @Override
  public final boolean unsafeCasMarker2V(WeakReference<Object> ov, WeakReference<Object> nv) {
    return MARKER_B.compareAndSet(this, ov, nv);
  }

  @Override
  public final WeakReference<Object> unsafeCmpxchgMarker2R(WeakReference<Object> ov, WeakReference<Object> nv) {
    return (WeakReference<Object>) MARKER_B.compareAndExchangeRelease(this, ov, nv);
  }

  @Override
  public final long id1() {
    return this._id1;
  }

  @Override
  public final long dummyImpl1(byte v) {
    return this.dummyImpl(v);
  }
}

File: core/native/src/main/scala/dev/tauri/choam/internal/refs/RefP1P1.scala
/*
 * SPDX-License-Identifier: Apache-2.0
 * Copyright 2016-2026 Daniel Urban and contributors listed in NOTICE.txt
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package dev.tauri.choam
package internal
package refs

import java.lang.ref.WeakReference

import scala.scalanative.annotation.alwaysinline

import mcas.Version
import core.Ref

private final class RefP1P1[A, B](
  a: A,
  b: B,
  i0: Long,
  i1: Long,
) extends PaddingForP1P1[A, B](a, i0) with Ref2Impl[A, B] {

  private[this] val _id1: Long =
    i1

  @volatile
  private[this] var valueB: B =
    b

  @volatile
  @nowarn("cat=unused-privates")
  private[this] var versionB: Long =
    Version.Start

  @volatile
  @nowarn("cat=unused-privates")
  private[this] var markerB: WeakReference[AnyRef] =
    _

  private[this] val refB: Ref[B] =
    new Ref2Ref2[A, B](this)

  @alwaysinline
  private[this] final def atomicValueB: AtomicHandle[B] = {
    AtomicHandle(this, "valueB")
  }

  @alwaysinline
  private[this] final def atomicVersionB: AtomicLongHandle = {
    AtomicLongHandle(this, "versionB")
  }

  @alwaysinline
  private[this] final def atomicMarkerB: AtomicHandle[WeakReference[AnyRef]] = {
    AtomicHandle(this, "markerB")
  }

  final override def _2: Ref[B] =
    this.refB

  protected[this] final override def refToString(): String = {
    refStringFrom2(this.id0(), this.id1())
  }

  final override def dummyImpl1(v: Byte): Long = {
    this.dummyImpl(v.toLong)
  }

  final override def id1(): Long =
    _id1

  final override def unsafeCas2V(ov: B, nv: B): Boolean = {
    atomicValueB.compareAndSet(ov, nv)
  }

  final override def unsafeCasMarker2V(ov: WeakReference[AnyRef], nv: WeakReference[AnyRef]): Boolean = {
    atomicMarkerB.compareAndSet(ov, nv)
  }

  final override def unsafeCmpxchg2R(ov: B, nv: B): B = {
    atomicValueB.compareAndExchangeRelAcq(ov, nv) // TODO: release-only
  }

  final override def unsafeCmpxchg2V(ov: B, nv: B): B = {
    atomicValueB.compareAndExchange(ov, nv)
  }

  final override def unsafeCmpxchgMarker2R(ov: WeakReference[AnyRef], nv: WeakReference[AnyRef]): WeakReference[AnyRef] = {
    atomicMarkerB.compareAndExchangeRelAcq(ov, nv) // TODO: release-only
  }

  final override def unsafeCmpxchgVersion2V(ov: Long, nv: Long): Long = {
    atomicVersionB.compareAndExchange(ov, nv)
  }

  final override def unsafeGet2P(): B = {
    atomicValueB.getOpaque // TODO: plain
  }

  final override def unsafeGet2V(): B = {
    this.valueB
  }

  final override def unsafeGetMarker2V(): WeakReference[AnyRef] = {
    this.markerB
  }

  final override def unsafeGetVersion2V(): Long = {
    this.versionB
  }

  final override def unsafeSet2P(nv: B): Unit = {
    atomicValueB.setOpaque(nv) // TODO: plain
  }

  final override def unsafeSet2V(nv: B): Unit = {
    this.valueB = nv
  }

  final override def dummyImpl2(v: Byte): Long =
    this.dummyImpl(v.toLong)
}
